board/nano-33-ble: pins, blinking leds and serial

ysoldak opened the pull request:
Support Arduino Nano 33 BLE [Sense] boards.

Flashed and verified with `blinky1`, `blinky2` and `echo` examples.

Requires special version of bossac, that can be easily obtained either via Arduino IDE or by simple download.
See comments in the header of the board definition. (thanks @deadprogram for finding the second option!)

~Examples changed a bit, hope that's OK.~

ysoldak committed:
board/nano-33-ble: pins, blinking leds and serial

ysoldak commented on src/examples/echo/echo.go:
```diff
@@ -25,7 +25,7 @@ func main() {
 
 			switch data {
-			case 13:
+			case 10, 13:
```
Makes it work on Mac

ysoldak commented on src/examples/blinky2/blinky2.go:
```diff
@@ -11,34 +11,18 @@ import (
 
-func led1() {
-	led := machine.LED1
+func led(led machine.Pin, interval time.Duration, padding string) {
```
Just merged functions in one. Makes it easier ad-hoc test more than 2 LEDs.

ysoldak commented on src/machine/board_nano-33-ble.go:
```diff
@@ -0,0 +1,80 @@
+
+package machine
+
+const HasLowFrequencyCrystal = false
```
I don't know if this board actually has it.
It works with this value, probably will work even better if enabled, I just don't know how to test.

aykevl replied:
If it works with `HasLowFrequencyCrystal = true`, there is probably a low frequency oscillator. Most boards (except for the smallest/cheapest) have one.

ysoldak commented on src/machine/board_nano-33-ble.go:
```diff
@@ -0,0 +1,80 @@
+// Onboard LEDs
+const (
+	LED         = LED_BUILTIN
+	LED1        = LED_RED
```
Numbered LEDs just to make it easier with the examples.

ysoldak commented on targets/nano-33-ble.ld:
```diff
@@ -0,0 +1,11 @@
+
+MEMORY
+{
+    FLASH_TEXT (rw) : ORIGIN = 0x10000, LENGTH = 1M
```
ORIGIN is important, hence separate .ld file

aykevl replied:
I see. However, this is not quite correct: the chip has only 1M of flash and `LENGTH = 1M` will go over this. It should be this:

```suggestion
    FLASH_TEXT (rw) : ORIGIN = 0x10000, LENGTH = 0xf0000
```

Unfortunately, this also means it's probably not possible to use the SoftDevice on this chip without destroying the bootloader.

ysoldak commented on targets/nano-33-ble.json:
```diff
@@ -0,0 +1,7 @@
+{
+	"inherits": ["nrf52840"],
+	"build-tags": ["nano_33_ble"],
+	"flash-command": "bossac_arduino2 -d -i -e -w -v -R --port={port} {bin}",
```
special version of bossac and no `--offset`

aykevl commented on targets/nano-33-ble.ld:
```diff
@@ -0,0 +1,11 @@
+    FLASH_TEXT (rw) : ORIGIN = 0x10000, LENGTH = 1M
+    RAM (rwx)       : ORIGIN = 0x20000000, LENGTH = 256K
+}
+OUTPUT_FORMAT ("elf32-littlearm", "elf32-bigarm", "elf32-littlearm")
```
I don't think this line is needed.

aykevl reviewed (commented):
Can you please separate the changes to the examples into a new PR? That makes it easier to discuss these changes.

ysoldak commented:
- Added docs to the board file.
- Removed examples from the PR.
- Adjusted `.ld` file according to suggestions, verified -- still works.
- Low frequency crystal boolean set to "true" and it still works, leaving it in this state.
- Squashed and rebased on current dev.

aykevl commented:
Thank you!
It's a bit unfortunate that this needs a special version of bossac but not much we can do about that :shrug:

deadprogram commented:
> Unfortunately, this also means it's probably not possible to use the SoftDevice on this chip without destroying the bootloader.

I expected that. 

There is UF2 bootloader available for this board https://github.com/adafruit/Adafruit_nRF52_Bootloader/tree/master/src/boards/arduino_nano_33_ble

However flashing the bootloader will require connecting to the debug pins on bottom of the board. Not too easy for most users.
